### DISCUSSION TITLE
Show & Tell: STDPOnlineLearner — STDP timing kernel and Phase 5 weight update pipeline

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#227] [Show and tell]
## What We Built

`STDPOnlineLearner` is the Phase 5 online learning component that applies Hebbian weight updates in real-time from spike events emitted by the neuromorphic module. It implements the classic **Spike-Timing-Dependent Plasticity** rule — the biologically-grounded learning algorithm where synapse strength changes based on the relative timing of pre- and post-synaptic spikes.

---

## Architecture Flow

```
NeuromorphicModule
    │ spike event (pre, post, Δt)
    ▼
STDPOnlineLearner.apply_weight_update()
    │
    ├─► STDP Kernel: Δw = A+ * exp(-Δt/τ+) [LTP if Δt>0]
    │                Δw = -A- * exp(Δt/τ-) [LTD if Δt<0]
    │
    ├─► Safety Gate: ||Δw|| / MAX_DELTA_NORM > 1.0? → reject
    │
    ├─► Weight Matrix Update (async, under lock)
    │
    ├─► Blackboard.write("phase5.stdp.last_update_ts", ...)
    │
    └─► Phase5MetricsExporter.observe_weight_delta(delta_w_norm)
```

---

### 1. STDP Kernel Parameters (A+=0.01, A-=0.012, τ+=20ms, τ-=20ms)
Slightly asymmetric to bias toward LTD, preventing runaway potentiation. A- > A+ means the net effect of random spike timing is mildly depressive — a well-known stability technique. All four parameters are configurable via the `STDPConfig` dataclass so researchers can experiment without touching core logic.

### 2. MAX_DELTA_NORM = 1.0
Inherited directly from the Issue #210 safety invariant. The normalized delta is computed as `abs(Δw) / MAX_DELTA_NORM`. Updates where this exceeds 1.0 are **rejected** before touching the weight matrix. Values approaching 1.0 (threshold: 0.9) trigger a Prometheus alert via `Phase5MetricsExporter`, giving operators early warning before a hard rejection.

### 3. Single `asyncio.Lock` per learner
Ensures atomic read-modify-write on the weight matrix. We deliberately **do not** hold the lock during subsequent Blackboard writes — this reduces lock contention at the cost of a tiny timestamp race (the `last_update_ts` key may briefly lag the actual write by one `await` cycle). This is acceptable because timestamps are best-effort diagnostic data, not safety-critical state.

### 4. `snapshot_weights()` returns a deep copy
The Phase 5 rollback manager (Issue #216) holds long-lived references to weight snapshots. Returning a view would allow silent mutation of the snapshot if the weight matrix is updated before rollback triggers. Deep copy eliminates this class of bug entirely at the cost of one allocation per hot-reload cycle (which is infrequent).

### 5. Module placement: `src/phase5/online_learning.py`
Keeps all Phase 5 components co-located for discoverability. `NeuromorphicBlackboardAdapter` imports `STDPOnlineLearner` from there, maintaining the adapter pattern established in Phase 4. Phase 5 code never reaches back into Phase 3/4 modules — dependency flows one way.

---

## STDP Timing Intuition

In biological neural networks:

- **Pre fires before post** (Δt > 0, "pre leads post") → synapse **strengthens** (**LTP** — Long-Term Potentiation). The pre-synaptic neuron was "predictive" of the post-synaptic response.
- **Post fires before pre** (Δt < 0) → synapse **weakens** (**LTD** — Long-Term Depression). The pre-synaptic neuron arrived too late to be causal.

This captures the Hebbian maxim: *"neurons that fire together, wire together"* — but with temporal precision. The exponential decay in both branches means the effect diminishes rapidly as |Δt| grows; spikes more than ~100ms apart have negligible weight impact.

---

## Integration Points

| Component | Interaction |
|---|---|
| `NeuromorphicBlackboardAdapter._on_spike_event()` | Calls `apply_weight_update()` on every qualifying spike pair |
| `Phase5RollbackManager` | Calls `snapshot_weights()` before each hot-reload cycle |
| `Phase5MetricsExporter` | Receives `observe_weight_delta(delta_w_norm)` for histogram tracking |
| `MeshCoordinator` | Gets `MODEL_HOT_RELOAD` broadcast after successful weight update (via Blackboard) |
| Blackboard | Key `phase5.stdp.last_update_ts` updated after each accepted delta |

---

## Open Question

Should the STDP kernel be **pluggable** (strategy pattern) to support future learning rule variants?

Candidates that have come up in discussions:
- **BCM rule** (Bienenstock-Cooper-Munro) — sliding threshold that adapts based on post-synaptic activity history
- **Oja's rule** — normalized Hebbian learning with built-in weight decay
- **Triplet STDP** — extends the pairwise model with triplet interactions for more biologically accurate dynamics

Currently the kernel is hardcoded. A strategy interface (`STDPKernelProtocol`) could make this pluggable with zero changes to `apply_weight_update()` internals.

See **Issue #181** for the full `STDPOnlineLearner` spec. Feedback welcome — especially from anyone who has worked with BCM dynamics in practice.
